File: leetcode/scala/198.house-robber.scala
object Solution {
  def rob(nums: Array[Int]): Int = {
    if (nums.length == 0) {
      return 0
    }
    if (nums.length == 1) {
      return nums(0)
    }
    if (nums.length == 2) return {
      if (nums(0) > nums(1)) nums(0)
      else nums(1)
    }
    val res = new Array[Int](nums.length)
    res(0) = nums(0)
    res(1) = if (nums(0) > nums(1)) nums(0) else nums(1)

    for (i <- Range(2, nums.length)) {
      res(i) = if (res(i - 1) > (res(i - 2) + nums(i))) {
        res(i - 1)
      }
      else {
        (res(i - 2) + nums(i))
      }
    }

    res(res.length - 1)
  }
}

File: leetcode/python/198.house-robber.py
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0
        if len(nums) == 1:
            return nums[0]
        dp: List[int] = [nums[0], max(nums[0], nums[1])]

        for i in range(2, len(nums)):
            dp.append(max(dp[i - 1], dp[i - 2] + nums[i]))

        return dp[-1]
